src/rules/valid-expect-in-promise.ts

```diff
@@ -0,0 +1,216 @@
+  if (body.type === AST_NODE_TYPES.BlockStatement) {
+    return body.body.find(line => {
+      if (line.type === AST_NODE_TYPES.ExpressionStatement) {
+        return isFullExpectCall(line.expression);
+      }
+      if (line.type === AST_NODE_TYPES.ReturnStatement && line.argument) {
+        return isFullExpectCall(line.argument);
+      }
+    });
+  }
+
+  return isFullExpectCall(body);
+};
+
+const isFullExpectCall = (expression: TSESTree.Node) =>
+  expression.type === AST_NODE_TYPES.CallExpression &&
+  isExpectMember(expression.callee) &&
+  isValidExpectCall(expression.callee.object);
+
+const reportReturnRequired = (context: RuleContext, node: TSESTree.Node) => {
+  context.report({
+    loc: {
+      end: {
+        column: node.loc.end.column,
+        line: node.loc.end.line,
+      },
+      start: node.loc.start,
+    },
+    messageId: 'returnPromise',
+    node,
+  });
+};
+
+const isPromiseReturnedLater = (
+  node: TSESTree.Node,
+  testFunctionBody: TSESTree.Statement[],
+) => {
+  let promiseName;
+  if (
+    node.type === AST_NODE_TYPES.ExpressionStatement &&
+    node.expression.type === AST_NODE_TYPES.CallExpression &&
+    node.expression.callee.type === AST_NODE_TYPES.MemberExpression &&
+    isSupportedAccessor(node.expression.callee.object)
+  ) {
+    promiseName = getAccessorValue(node.expression.callee.object);
+  } else if (
+    node.type === AST_NODE_TYPES.VariableDeclarator &&
+    node.id.type === AST_NODE_TYPES.Identifier
+  ) {
+    promiseName = node.id.name;
+  }
+
+  const lastLineInTestFunc = testFunctionBody[testFunctionBody.length - 1];
+
+  return (
+    lastLineInTestFunc.type === AST_NODE_TYPES.ReturnStatement &&
+    lastLineInTestFunc.argument &&
+    (('name' in lastLineInTestFunc.argument &&
+      lastLineInTestFunc.argument.name === promiseName) ||
+      !promiseName)
+  );
+};
+
+const isTestFunc = (node: TSESTree.Node) =>
+  node.type === AST_NODE_TYPES.CallExpression &&
+  isSupportedAccessor(node.callee) &&
+  ([TestCaseName.it, TestCaseName.test] as string[]).includes(
+    getAccessorValue(node.callee),
+  );
+
+const findTestFunction = (node: TSESTree.Node | undefined) => {
+  while (node) {
+    if (isFunction(node) && node.parent && isTestFunc(node.parent)) {
+      return node;
+    }
+
+    node = node.parent;
+  }
+
+  return null;
+};
+
+const isParentThenOrPromiseReturned = (
+  node: TSESTree.Node,
+  testFunctionBody: TSESTree.Statement[],
+) =>
+  node.type === AST_NODE_TYPES.ReturnStatement ||
+  isPromiseReturnedLater(node, testFunctionBody);
+
+// prettier-ignore
+// WEB-40105
+type PromiseCallbacks = [
+  TSESTree.Expression | undefined,
+  TSESTree.Expression | undefined
+]
+
+const verifyExpectWithReturn = (
+  promiseCallbacks: PromiseCallbacks,
+  node: CalledKnownMemberExpression<'then' | 'catch'>,
+  context: RuleContext,
+  testFunctionBody: TSESTree.Statement[],
+) => {
+  promiseCallbacks.some(promiseCallback => {
```

Comment:
this isn't `return`ed, why use `.some`? To short-circuit? If so, please use `find`

Reply:
No, b/c this [calls `reportReturnRequired`](https://github.com/jest-community/eslint-plugin-jest/blob/03a9663a05bcd10781cd44f55bab7c2745655a1a/src/rules/valid-expect-in-promise.ts#L147):

https://github.com/jest-community/eslint-plugin-jest/blob/03a9663a05bcd10781cd44f55bab7c2745655a1a/src/rules/valid-expect-in-promise.ts#L52-L64

You could replace it w/ find, and then do a check on that.. which yeah I think would look nicer.

Reply:
Ah no, if we use `find`, then `node.parent.parent` is not within the guard, so we'd have to double the defined check.

I think it's just simpler to leave this for now. I want to try and clean this code up separately at some point anyway to try and reduce the number of checks it + `valid-expect` do.